eth/downloader/queue.go

```diff
@@ -349,66 +341,78 @@ func (q *queue) Schedule(headers []*types.Header, from uint64) []*types.Header {
 
 // Results retrieves and permanently removes a batch of fetch results from
 // the cache. the result slice will be empty if the queue has been closed.
+// This is 'thread-safe', but assumes that there are not two simultaneous
+// callers to Results (both will modify q.resultSize)
 func (q *queue) Results(block bool) []*fetchResult {
-	q.lock.Lock()
-	defer q.lock.Unlock()
 
-	// Count the number of items available for processing
-	nproc := q.countProcessableItems()
-	for nproc == 0 && !q.closed {
-		if !block {
-			return nil
+	// abort early if there are no items and non-blocking requested
+	if !q.resultCache.HasCompletedItems() && !block {
+		return nil
+	}
+	results := q.resultCache.GetCompleted(maxResultsProcess)
+	if len(results) == 0 && !block {
+		return nil
+	}
+	closed := false
+	for !closed && len(results) == 0 {
+		// In order to wait on 'active', we need to obtain the lock.
+		// That may take a while, if someone is delivering at the same
+		// time, so after obtaining the lock, we check again if there
+		// are any results to fetch.
+		// Also, in-between we ask for the lock and the lock is obtained,
+		// someone can have closed the queue. In that case, we should
+		// return the available results and stop blocking
+		q.lock.Lock()
+		closed = q.closed
+		results = q.resultCache.GetCompleted(maxResultsProcess)
+		if closed || len(results) > 0 {
+			q.lock.Unlock()
+			break
 		}
 		q.active.Wait()
-		nproc = q.countProcessableItems()
-	}
-	// Since we have a batch limit, don't pull more into "dangling" memory
-	if nproc > maxResultsProcess {
-		nproc = maxResultsProcess
-	}
-	results := make([]*fetchResult, nproc)
-	copy(results, q.resultCache[:nproc])
-	if len(results) > 0 {
-		// Mark results as done before dropping them from the cache.
-		for _, result := range results {
-			hash := result.Header.Hash()
-			delete(q.blockDonePool, hash)
-			delete(q.receiptDonePool, hash)
+		closed = q.closed
+		q.lock.Unlock()
+		results = q.resultCache.GetCompleted(maxResultsProcess)
```

Comment:
I think the condition variable should be in `resultStore`. This means `closed` needs to move into the `resultStore` as well.

Reply:
Well, that totally makes sense, but it means an even larger refactor. Then `closed` would have to move in there, and the things that calls `Signal` need to somehow trigger that via the `resultStore`. 
Let's leave that for a future refactor (I'd be happy to continue iterating on the downloader)